NorthstarDLL/server/auth/serverauthentication.cpp

```diff
@@ -229,51 +227,72 @@ void*,, (
-bool,, (R2::CBaseClient* self, char* pName, void* pNetChannel, char bFakePlayer, void* a5, char pDisconnectReason[256], void* a7))
-// clang-format on
-{
 	const char* pAuthenticationFailure = nullptr;
 	char pVerifiedName[64];
+	const char* authToken = serverFilter;
 
-	if (!bFakePlayer)
+	if (!client->m_bFakePlayer)
 	{
-		if (!g_pServerAuthentication->VerifyPlayerName(pNextPlayerToken, pName, pVerifiedName))
+		if (!g_pServerAuthentication->VerifyPlayerName(authToken, playerName, pVerifiedName))
 			pAuthenticationFailure = "Invalid Name.";
-		else if (!g_pBanSystem->IsUIDAllowed(iNextPlayerUid))
+		else if (!g_pBanSystem->IsUIDAllowed(uid))
 			pAuthenticationFailure = "Banned From server.";
-		else if (!g_pServerAuthentication->CheckAuthentication(self, iNextPlayerUid, pNextPlayerToken))
+		else if (!g_pServerAuthentication->CheckAuthentication(client, uid, authToken))
 			pAuthenticationFailure = "Authentication Failed.";
 	}
-	else // need to copy name for bots still
-		strncpy_s(pVerifiedName, pName, 63);
+	else
+	{
+		spdlog::error("A bot called CBaseServer__ConnectClient, should be impossible!");
+
+		CBaseServer__PushDisconnectReason(self, (int32_t)((uintptr_t)self + 0xc), addr, "A bot was init in CServer__ConnectClient");
+		return nullptr;
+	}
 
 	if (pAuthenticationFailure)
 	{
-		spdlog::info("{}'s (uid {}) connection was rejected: \"{}\"", pName, iNextPlayerUid, pAuthenticationFailure);
+		spdlog::info("{}'s (uid {}) connection was rejected: \"{}\"", playerName, uid, pAuthenticationFailure);
 
-		strncpy_s(pDisconnectReason, 256, pAuthenticationFailure, 255);
-		return false;
+		CBaseServer__PushDisconnectReason(self, (int32_t)((uintptr_t)self + 0xc), addr, pAuthenticationFailure);
+		return nullptr;
 	}
 
-	// try to actually connect the player
-	if (!CBaseClient__Connect(self, pVerifiedName, pNetChannel, bFakePlayer, a5, pDisconnectReason, a7))
+	// write name into the client
+	strncpy_s(pVerifiedName, client->m_Name, 63);
+
+	// we already know this player's authentication data is legit, actually write it to them now
+	g_pServerAuthentication->AuthenticatePlayer(client, uid, authToken);
+
+	g_pServerAuthentication->AddPlayer(client, authToken);
+	g_pServerLimits->AddPlayer(client);
+}
+
+ConVar* Cvar_ns_allowuserclantags;
+
+// clang-format off
+AUTOHOOK(CBaseClient__Connect, engine.dll + 0x101740,
+bool,, (R2::CBaseClient* self, char* pName, void* pNetChannel, char bFakePlayer, void* a5, char pDisconnectReason[256], void* a7))
+// clang-format on
+{
+	// only remains to count bots in player count,
+	// since bots take player slots and it will give if not counted a false player count on the server browser.
+
+	if (!bFakePlayer)
+		return CBaseClient__Connect(self, pName, pNetChannel, bFakePlayer, a5, pDisconnectReason, a7);
+
+	// try to actually connect the bot
+	else if (!CBaseClient__Connect(self, pName, pNetChannel, bFakePlayer, a5, pDisconnectReason, a7))
 		return false;
```

Comment:
This control flow could be improved a bit I think? (At the very least the `else` is unneeded since the above `if` always returns)